File: Java/src/com/dbc/algorithm/UnionFindSets/RemoveMaxNumberOfEdgesToKeepGraphFullyTraversable.java
package com.dbc.algorithm.UnionFindSets;

import java.util.ArrayList;
import java.util.List;

public class RemoveMaxNumberOfEdgesToKeepGraphFullyTraversable {
    public static int find(int[] parent, int index) {
        if (index != parent[index]) {
            parent[index] = find(parent, parent[index]);
        }
        return parent[index];
    }

    public static void union(int[] parent, int index1, int index2) {
        parent[find(parent, index2)] = find(parent, index1);
    }

    public static int maxNumEdgesToRemove(int n, int[][] edges) {
        int[] parent = new int[n + 1];
        for (int i = 0; i < n + 1; i++) {
            parent[i] = i;
        }

        int count = 0;
        int flag = 0;

        List<int[]> alice = new ArrayList<>();
        List<int[]> bob = new ArrayList<>();
        for (int[] edge : edges) {
            if (edge[0] == 1)
                alice.add(edge);
            else if (edge[0] == 2)
                bob.add(edge);
            else {
                int u = find(parent, edge[1]);
                int v = find(parent, edge[2]);
                if (u == v)
                    count++;
                else {
                    flag++;
                    union(parent, edge[1], edge[2]);
                }
            }
        }

        int temp = flag;
        int[] alice_parent = parent.clone();
        int[] bob_parent = parent.clone();
        for (int[] edge : alice) {
            int u = find(alice_parent, edge[1]);
            int v = find(alice_parent, edge[2]);
            if (u == v)
                count++;
            else {
                temp++;
                union(alice_parent, edge[1], edge[2]);
            }
        }

        if (temp < n - 1)
            return -1;
        temp = flag;
        for (int[] edge : bob) {
            int u = find(bob_parent, edge[1]);
            int v = find(bob_parent, edge[2]);
            if (u == v)
                count++;
            else {
                temp++;
                union(bob_parent, edge[1], edge[2]);
            }
        }

        if (temp < n - 1)
            return -1;
        return count;
    }
}

File: src/tmp/Python/union-find-sets/remove_max_number_of_edges_to_keep_graph_fully_traversable.py
class UnionFind():
    def __init__(self, parent: list) -> None:
        self.parent = parent

    def find(self, index: int) -> int:
        if index != self.parent[index]:
            self.parent[index] = self.find(self.parent[index])
        return self.parent[index]

    def union(self, index1: int, index2) -> None:
        self.parent[self.find(index2)] = self.find(index1)


def maxNumEdgesToRemove(n: int, edges: list) -> int:
    parent = list(range(n+1))
    alice = []
    bob = []
    count = 0
    flag = 0

    type1 = UnionFind(parent)
    for edge in edges:
        if edge[0] == 1:
            alice.append(edge)
        elif edge[0] == 2:
            bob.append(edge)
        else:
            u = type1.find(edge[1])
            v = type1.find(edge[2])
            if u == v:
                count += 1
            else:
                type1.union(u, v)
                flag += 1

    temp = flag
    type2 = UnionFind(parent.copy())
    type3 = UnionFind(parent.copy())
    for edge in alice:
        u = type2.find(edge[1])
        v = type2.find(edge[2])
        if u == v:
            count += 1
        else:
            type2.union(u, v)
            temp += 1

    if temp < n - 1:
        return -1
    temp = flag
    for edge in bob:
        u = type3.find(edge[1])
        v = type3.find(edge[2])
        if u == v:
            count += 1
        else:
            type3.union(u, v)
            temp += 1

    if temp < n - 1:
        return -1

    return count


if __name__ == "__main__":
    print(maxNumEdgesToRemove(n = 4, edges = [[3,2,3],[1,1,2],[2,3,4]]))
